src/mds/Server.cc

```diff
@@ -5575,26 +5576,63 @@ void Server::handle_set_vxattr(MDRequestRef& mdr, CInode *cur)
     }
 
     if (quota.is_enable() && !cur->get_projected_srnode())
-      new_realm = true;
+      adjust_realm = true;
 
-    if (!xlock_policylock(mdr, cur, false, new_realm))
+    if (!xlock_policylock(mdr, cur, false, adjust_realm))
       return;
 
-    auto pi = cur->project_inode(mdr, false, new_realm);
+    if (cur->get_projected_inode()->quota == quota) {
+      respond_to_request(mdr, 0);
+      return;
+    }
+
+    auto pi = cur->project_inode(mdr, false, adjust_realm);
     pi.inode->quota = quota;
 
-    if (new_realm) {
-      SnapRealm *realm = cur->find_snaprealm();
-      auto seq = realm->get_newest_seq();
-      auto &newsnap = *pi.snapnode;
-      newsnap.created = seq;
-      newsnap.seq = seq;
-    }
+    if (adjust_realm)
+      pi.snapnode->created = pi.snapnode->seq = cur->find_snaprealm()->get_newest_seq();
+
     mdr->no_early_reply = true;
     pip = pi.inode.get();
 
     client_t exclude_ct = mdr->get_client();
     mdcache->broadcast_quota_to_client(cur, exclude_ct, true);
+  } else if (name == "ceph.dir.subvolume"sv) {
+    if (!cur->is_dir() || cur->is_root() ||
+	!cur->get_projected_parent_dir()->inode->is_root()) { // must in root dir
```

Comment:
this won't work for pybind/mgr/volumes. the directories will be in:

`/volumes/<group>/<volume name>/<uuid>`

Reply:
if we don't limit location of subvolume. we may have nested subvolumes, which will make subvolume useless.

Reply:
Well then there needs to be nested subvolume detection. This can't be changed in the mgr/volumes plugin.

Reply:
there is no easy way to detect child subvolume

Reply:
> there is no easy way to detect child subvolume

If it is truly difficult there are ways around. We can add a mark for a "subvolume root" on the subvolume parent to be checked instead. Trust the admin (mgr/volumes) to do the right thing configuring that. Keep in mind there would be multiple subvolume roots managed by mgr/volumes but no overlap.